examples/fluids/src/dirichlet.c

```diff
@@ -0,0 +1,194 @@
+  PetscFunctionBeginUser;
+
+  DMLabel domain_label;
+  PetscCall(DMGetLabel(dm, "Face Sets", &domain_label));
+
+  // Basis
+  CeedInt height = 1;
+  PetscCall(CeedBasisCreateProjection(ceed_data->basis_x_sur,
+                                      ceed_data->basis_q_sur, &basis_x_to_q_sur));
+
+  // Setup QFunction
+  CeedQFunctionCreateInterior(ceed, 1, SetupDirichletBC, SetupDirichletBC_loc,
+                              &qf_setup);
+  CeedQFunctionAddInput(qf_setup, "x", num_comp_x, CEED_EVAL_INTERP);
+  CeedQFunctionAddInput(qf_setup, "multiplicity", num_comp_q, CEED_EVAL_NONE);
+  CeedQFunctionAddOutput(qf_setup, "x stored", num_comp_x, CEED_EVAL_NONE);
+  CeedQFunctionAddOutput(qf_setup, "scale", 1, CEED_EVAL_NONE);
+
+  // Compute contribution on each boundary face
+  for (CeedInt i=0; i < bc->num_inflow; i++) {
+    // -- Restrictions
+    PetscCall(GetRestrictionForDomain(ceed, dm, height, domain_label,
+                                      bc->inflows[i],
+                                      Q_sur, q_data_size_sur, &elem_restr_q_sur, &elem_restr_x_sur,
+                                      &elem_restr_qd_sur));
+    CeedElemRestrictionCreateVector(elem_restr_q_sur, &multiplicity, NULL);
+    CeedElemRestrictionGetMultiplicity(elem_restr_q_sur, multiplicity);
+    CeedElemRestrictionGetNumElements(elem_restr_q_sur, &num_elem);
+    CeedElemRestrictionGetElementSize(elem_restr_q_sur, &elem_size);
+    CeedElemRestrictionCreateStrided(ceed, num_elem, elem_size, num_comp_x,
+                                     num_elem * elem_size * num_comp_x,
+                                     CEED_STRIDES_BACKEND, &elem_restr_x_stored);
+    CeedElemRestrictionCreateVector(elem_restr_x_stored, &x_stored, NULL);
+
+    CeedElemRestrictionCreateStrided(ceed, num_elem, elem_size, 1,
+                                     num_elem * elem_size,
+                                     CEED_STRIDES_BACKEND, &elem_restr_scale);
+    CeedElemRestrictionCreateVector(elem_restr_scale, &scale_stored, NULL);
+
+    CeedVectorCreate(ceed, q_data_size_sur*num_elem*elem_size, &q_data_sur);
+
+    // -- Setup Operator
+    CeedOperatorCreate(ceed, qf_setup, NULL, NULL, &op_setup);
+    CeedOperatorSetName(op_setup, "surface geometric data");
+    CeedOperatorSetField(op_setup, "x", elem_restr_x_sur, basis_x_to_q_sur,
+                         CEED_VECTOR_ACTIVE);
+    CeedOperatorSetField(op_setup, "multiplicity", elem_restr_q_sur,
+                         CEED_BASIS_COLLOCATED, multiplicity);
+    CeedOperatorSetField(op_setup, "x stored", elem_restr_x_stored,
+                         CEED_BASIS_COLLOCATED, x_stored);
+    CeedOperatorSetField(op_setup, "scale", elem_restr_scale, CEED_BASIS_COLLOCATED,
+                         scale_stored);
+
+    // -- Compute geometric factors
+    CeedOperatorApply(op_setup, ceed_data->x_coord, CEED_VECTOR_NONE,
+                      CEED_REQUEST_IMMEDIATE);
+
+    // -- Compute QData for the surface
```

Comment:
I'm not sure I follow. The only input or output that's at nodes is `ceed_data->x_coord`. Everything else is set as `CEED_BASIS_COLLOCATED`. And for the dirichlet routines we've set the quadrature points to be at the nodes.

Reply:
You are computing qdata at the quadratre points and using it at the nodes

Reply:
`basis_x_sur` goes from the coordinates to the quadratre space, not to the nodal space

Reply:
Ohhhh. I thought you were talking at the `op_setup` code, not the `op_setup_sur` code. Yeah, poor copy/paste. 

Should be `basis_x_to_q_sur` for `"dx"` and `"weight"` right? Granted we don't use anything from the quadrature weight in this case, so that one probably doesn't matter in this case.

Reply:
The problem is that gradients are not defined for the projection basis

Reply:
Hmmm. Has it just not been implemented, or is that a permanent issue?

In the meantime, I can replace it with a "normal" basis that goes from points to nodes. Like:

```c
  CeedBasisCreateTensorH1Lagrange(ceed, dim, num_comp_x, 2, P,
                                  CEED_GAUSS_LOBATTO, &ceed_data->basis_xc_sur);
```

Reply:
I am not sure of an obvious way to define the gradient of the projection with what we have in each basis?

Reply:
We have $B_p = B_f^\dagger B_c$ and I think $D_p = B_F^\dagger D_c$ is the gradient, in the sense that $B_f D_p \approx D_c$ (exactly when c is a subset of the space f).

Reply:
I'll resolve this thread because it isn't really part of this PR, but feel free to make it into an issue.